File: swift/Engine/Sources/SorobanEngine/Sheet/Controls.swift
import Anzan
/// What-if sliders, the first CONTROL EXPRESSION: a cell whose expression is
/// a literal-argument `slider(value, min, max[, step])` call renders as a
/// draggable control, and dragging rewrites the value literal in place.
/// Combined with 𝑖 definitions (`rate = slider(0.08, 0, 0.2)`) the control
/// is named, sheet-scoped, and immutable from the log — all inherited.
///
/// The pattern is deliberately general: checkbox/stepper/dropdown can follow
/// it later (a control call with a literal "storage" argument, rewritten by
/// interaction).

/// Everything the grid needs to draw and drag one slider.
public struct SliderInfo: Equatable, Sendable {
    /// The 𝑖 name when the slider is a definition; nil for an anonymous
    /// `=slider(…)` cell (read it by address instead).
    public let name: String?
    /// Clamped into minimum...maximum.
    public let value: BigDecimal
    public let minimum: BigDecimal
    public let maximum: BigDecimal
    /// Explicit 4th argument, or (max−min)/100 — an exact exponent shift.
    public let step: BigDecimal

    /// Builds from a `slider(…)`/`stepper(…)` call whose arguments are all
    /// numeric LITERALS (the value argument IS the storage — it can't be an
    /// expression). Returns nil for any other shape; invalid ranges fall
    /// through to normal evaluation, which reports the error.
    /// Default step: (max−min)/100 for sliders, 1 for steppers.
    static func extract(from expression: Expression, name: String?,
                        function: String = "slider") -> SliderInfo? {
        guard case .call(let callName, let arguments) = expression,
              callName.lowercased() == function,
              (3...4).contains(arguments.count) else { return nil }

        var literals: [BigDecimal] = []
        for argument in arguments {
            guard case .number(let literal)? = Control.literalValue(argument) else { return nil }
            literals.append(literal)
        }
        let minimum = literals[1], maximum = literals[2]
        guard minimum < maximum else { return nil } // evaluation reports this
        if literals.count == 4, !(literals[3] > .zero) { return nil }

        let span = maximum - minimum
        let step: BigDecimal
        if literals.count == 4 {
            step = literals[3]
        } else if function == "stepper" {
            step = .one
        } else {
            step = BigDecimal(significand: span.significand, exponent: span.exponent - 2)
        }
        return SliderInfo(name: name,
                          value: min(max(literals[0], minimum), maximum),
                          minimum: minimum, maximum: maximum, step: step)
    }

    /// The value `fraction` (0...1) of the way along the track, quantized to
    /// `step` (from the minimum) and clamped — what a drag position means.
    public func value(atFraction fraction: Double) -> BigDecimal {
        let clamped = Swift.min(Swift.max(fraction, 0), 1)
        let span = maximum - minimum
        // steps = round(fraction × span / step); Double is fine HERE because
        // it only picks which step — the returned value is exact step math.
        guard let spanValue = Double(span.description),
              let stepValue = Double(step.description), stepValue > 0 else { return minimum }
        let steps = (clamped * spanValue / stepValue).rounded()
        let candidate = minimum + step * BigDecimal(Int(steps))
        return Swift.min(Swift.max(candidate, minimum), maximum)
    }

    /// Where the knob sits, 0...1.
    public var fraction: Double {
        let span = maximum - minimum
        guard let spanValue = Double(span.description), spanValue > 0,
              let offset = Double((value - minimum).description) else { return 0 }
        return Swift.min(Swift.max(offset / spanValue, 0), 1)
    }

    /// The widest value label this slider can show in `format` — the grid
    /// reserves this width so the track never resizes mid-drag. The worst case
    /// of integer digits × step decimals is one of min/max and their one-step
    /// neighbors; the current value covers a long resting literal.
    public func widestValueText(format: NumberFormat) -> String {
        [minimum, minimum + step, maximum - step, maximum, value]
            .map(format.rendered).max { $0.count < $1.count } ?? ""
    }
}

/// A checkbox cell: `flag = checkbox(true)`. Clicking flips the literal.
public struct CheckboxInfo: Equatable, Sendable {
    public let name: String?
    public let isOn: Bool

    static func extract(from expression: Expression, name: String?) -> CheckboxInfo? {
        guard case .call(let callName, let arguments) = expression,
              callName.lowercased() == "checkbox", arguments.count == 1,
              case .number(let state)? = Control.literalValue(arguments[0]) else { return nil }
        return CheckboxInfo(name: name, isOn: !state.isZero)
    }
}

/// A dropdown cell: `region = dropdown("EU", ["EU", "US", "APAC"])`. The
/// cell's value IS the selected option; choosing rewrites the literal.
/// Options are literals too — strings or numbers.
public struct DropdownInfo: Equatable, Sendable {
    public let name: String?
    public let value: Value
    public let options: [Value]

    static func extract(from expression: Expression, name: String?) -> DropdownInfo? {
        guard case .call(let callName, let arguments) = expression,
              callName.lowercased() == "dropdown", arguments.count == 2,
              let value = Control.literalValue(arguments[0]),
              case .arrayLiteral(let items) = arguments[1], !items.isEmpty else { return nil }
        var options: [Value] = []
        for item in items {
            guard let option = Control.literalValue(item) else { return nil }
            options.append(option)
        }
        return DropdownInfo(name: name, value: value, options: options)
    }
}

public enum Control {
    static let names: Set<String> = ["slider", "stepper", "checkbox", "dropdown"]

    /// The literal forms a control's storage argument may take: numbers
    /// (optionally signed), true/false, and "strings".
    static func literalValue(_ expression: Expression) -> Value? {
        switch expression {
        case .number(let value): return .number(value)
        case .unaryMinus(.number(let value)): return .number(-value)
        case .stringLiteral(let text): return .string(text)
        case .variable(let name) where name.lowercased() == "true": return .number(.one)
        case .variable(let name) where name.lowercased() == "false": return .number(.zero)
        default: return nil
        }
    }

    /// The cell's control, if its content is a control expression: either a
    /// 𝑖 definition whose body is a control call (named) or a plain/`=`
    /// formula that IS one (anonymous).
    static func display(for cell: Cell) -> CellDisplay? {
        let expression: Expression
        let name: String?
        switch cell.content {
        case .definition(let definition):
            guard case .variable(let body) = definition.kind else { return nil }
            expression = body
            name = definition.name
        case .explicitFormula(.success(let body)), .candidate(let body):
            expression = body
            name = nil
        default:
            return nil
        }

        guard case .call(let callName, _) = expression else { return nil }
        switch callName.lowercased() {
        case "slider":
            return SliderInfo.extract(from: expression, name: name).map(CellDisplay.slider)
        case "stepper":
            return SliderInfo.extract(from: expression, name: name, function: "stepper")
                .map(CellDisplay.stepper)
        case "checkbox":
            return CheckboxInfo.extract(from: expression, name: name).map(CellDisplay.checkbox)
        case "dropdown":
            return DropdownInfo.extract(from: expression, name: name).map(CellDisplay.dropdown)
        default:
            return nil
        }
    }

    /// Rewrites a control's STORAGE argument literal inside the raw cell
    /// text, leaving everything else — spacing, the 𝑖 name, trailing
    /// `# comments` — intact. `literal` is the replacement source text
    /// (`0.11`, `true`, `"US"`). Token-precise via the lexer's ranges.
    public static func rewriting(_ raw: String, toLiteral literal: String) -> String? {
        guard let tokens = try? Lexer.tokenize(raw) else { return nil }
        for (index, token) in tokens.enumerated() {
            guard case .identifier(let name) = token.kind, names.contains(name.lowercased()),
                  index + 2 < tokens.count,
                  case .leftParen = tokens[index + 1].kind else { continue }

            var start = index + 2
            let range: Range<Int>
            switch tokens[start].kind {
            case .number, .string:
                range = tokens[start].range
            case .identifier(let word) where ["true", "false"].contains(word.lowercased()):
                range = tokens[start].range
            case .minus:
                start += 1
                guard start < tokens.count, case .number = tokens[start].kind else { return nil }
                range = tokens[start - 1].range.lowerBound..<tokens[start].range.upperBound
            default:
                return nil
            }

            var characters = Array(raw)
            characters.replaceSubrange(range, with: literal)
            return String(characters)
        }
        return nil
    }
}

public enum Slider {
    /// Numeric convenience kept for slider drags (and their tests).
    public static func rewriting(_ raw: String, to newValue: BigDecimal) -> String? {
        Control.rewriting(raw, toLiteral: newValue.description)
    }
}

File: rust/engine/src/controls.rs
//! Control expressions — sliders, steppers, checkboxes, dropdowns: a cell
//! whose expression is a literal-argument control call renders as an
//! interactive control, and interaction rewrites the storage literal in
//! place. Combined with 𝑖 definitions (`rate = slider(0.08, 0, 0.2)`) the
//! control is named, sheet-scoped, and immutable from the log — all
//! inherited.

use crate::cell::{Cell, Content, DefinitionKind};
use crate::spreadsheet::CellDisplay;
use anzan::ast::Expression;
use anzan::lexer::{Lexer, TokenKind};
use anzan::{BigDecimal, Value};

/// Everything the grid needs to draw and drag one slider (or stepper).
#[derive(Debug, Clone, PartialEq)]
pub struct SliderInfo {
    /// The 𝑖 name when the slider is a definition; `None` for an anonymous
    /// `=slider(…)` cell (read it by address instead).
    pub name: Option<String>,
    /// Clamped into minimum...maximum.
    pub value: BigDecimal,
    pub minimum: BigDecimal,
    pub maximum: BigDecimal,
    /// Explicit 4th argument, or (max−min)/100 — an exact exponent shift.
    pub step: BigDecimal,
}

impl SliderInfo {
    /// Builds from a `slider(…)`/`stepper(…)` call whose arguments are all
    /// numeric LITERALS (the value argument IS the storage — it can't be an
    /// expression). `None` for any other shape; invalid ranges fall through
    /// to normal evaluation, which reports the error. Default step:
    /// (max−min)/100 for sliders, 1 for steppers.
    pub(crate) fn extract(
        expression: &Expression,
        name: Option<&str>,
        function: &str,
    ) -> Option<SliderInfo> {
        let Expression::Call {
            name: call_name,
            arguments,
        } = expression
        else {
            return None;
        };
        if !call_name.eq_ignore_ascii_case(function) || !(3..=4).contains(&arguments.len()) {
            return None;
        }

        let mut literals: Vec<BigDecimal> = Vec::with_capacity(arguments.len());
        for argument in arguments {
            let Some(Value::Number(literal)) = Control::literal_value(argument) else {
                return None;
            };
            literals.push(literal);
        }
        let minimum = literals[1].clone();
        let maximum = literals[2].clone();
        if minimum >= maximum {
            return None; // evaluation reports this
        }
        if literals.len() == 4 && literals[3] <= BigDecimal::zero() {
            return None;
        }

        let span = &maximum - &minimum;
        let step = if literals.len() == 4 {
            literals[3].clone()
        } else if function == "stepper" {
            BigDecimal::one()
        } else {
            BigDecimal::new(span.significand().clone(), span.exponent() - 2)
        };
        let value = literals[0]
            .clone()
            .max(minimum.clone())
            .min(maximum.clone());
        Some(SliderInfo {
            name: name.map(str::to_string),
            value,
            minimum,
            maximum,
            step,
        })
    }
}

/// A checkbox cell: `flag = checkbox(true)`. Clicking flips the literal.
#[derive(Debug, Clone, PartialEq)]
pub struct CheckboxInfo {
    pub name: Option<String>,
    pub is_on: bool,
}

impl CheckboxInfo {
    fn extract(expression: &Expression, name: Option<&str>) -> Option<CheckboxInfo> {
        let Expression::Call {
            name: call_name,
            arguments,
        } = expression
        else {
            return None;
        };
        if !call_name.eq_ignore_ascii_case("checkbox") || arguments.len() != 1 {
            return None;
        }
        let Some(Value::Number(state)) = Control::literal_value(&arguments[0]) else {
            return None;
        };
        Some(CheckboxInfo {
            name: name.map(str::to_string),
            is_on: !state.is_zero(),
        })
    }
}

/// A dropdown cell: `region = dropdown("EU", ["EU", "US", "APAC"])`. The
/// cell's value IS the selected option; choosing rewrites the literal.
/// Options are literals too — strings or numbers.
#[derive(Debug, Clone, PartialEq)]
pub struct DropdownInfo {
    pub name: Option<String>,
    pub value: Value,
    pub options: Vec<Value>,
}

impl DropdownInfo {
    fn extract(expression: &Expression, name: Option<&str>) -> Option<DropdownInfo> {
        let Expression::Call {
            name: call_name,
            arguments,
        } = expression
        else {
            return None;
        };
        if !call_name.eq_ignore_ascii_case("dropdown") || arguments.len() != 2 {
            return None;
        }
        let value = Control::literal_value(&arguments[0])?;
        let Expression::ArrayLiteral(items) = &arguments[1] else {
            return None;
        };
        if items.is_empty() {
            return None;
        }
        let mut options = Vec::with_capacity(items.len());
        for item in items {
            options.push(Control::literal_value(item)?);
        }
        Some(DropdownInfo {
            name: name.map(str::to_string),
            value,
            options,
        })
    }
}

pub struct Control;

impl Control {
    const NAMES: [&'static str; 4] = ["slider", "stepper", "checkbox", "dropdown"];

    /// The literal forms a control's storage argument may take: numbers
    /// (optionally signed), true/false, and "strings".
    fn literal_value(expression: &Expression) -> Option<Value> {
        match expression {
            Expression::Number(value) => Some(Value::Number(value.clone())),
            Expression::UnaryMinus(inner) => {
                if let Expression::Number(value) = inner.as_ref() {
                    Some(Value::Number(-value))
                } else {
                    None
                }
            }
            Expression::StringLiteral(text) => Some(Value::String(text.clone())),
            Expression::Variable(name) if name.eq_ignore_ascii_case("true") => {
                Some(Value::Number(BigDecimal::one()))
            }
            Expression::Variable(name) if name.eq_ignore_ascii_case("false") => {
                Some(Value::Number(BigDecimal::zero()))
            }
            _ => None,
        }
    }

    /// The cell's control, if its content is a control expression: either a
    /// 𝑖 definition whose body is a control call (named) or a plain/`=`
    /// formula that IS one (anonymous).
    pub(crate) fn display(cell: &Cell) -> Option<CellDisplay> {
        let (expression, name): (&Expression, Option<&str>) = match &cell.content {
            Content::Definition(definition) => {
                let DefinitionKind::Variable(body) = &definition.kind else {
                    return None;
                };
                (body, Some(definition.name.as_str()))
            }
            Content::ExplicitFormula(Ok(body)) | Content::Candidate(body) => (body, None),
            _ => return None,
        };

        let Expression::Call {
            name: call_name, ..
        } = expression
        else {
            return None;
        };
        match call_name.to_lowercase().as_str() {
            "slider" => SliderInfo::extract(expression, name, "slider").map(CellDisplay::Slider),
            "stepper" => SliderInfo::extract(expression, name, "stepper").map(CellDisplay::Stepper),
            "checkbox" => CheckboxInfo::extract(expression, name).map(CellDisplay::Checkbox),
            "dropdown" => DropdownInfo::extract(expression, name).map(CellDisplay::Dropdown),
            _ => None,
        }
    }

    /// Rewrites a control's STORAGE argument literal inside the raw cell
    /// text, leaving everything else — spacing, the 𝑖 name, trailing
    /// `# comments` — intact. `literal` is the replacement source text
    /// (`0.11`, `true`, `"US"`). Token-precise via the lexer's ranges.
    pub fn rewriting(raw: &str, literal: &str) -> Option<String> {
        let tokens = Lexer::tokenize(raw).ok()?;
        for (index, token) in tokens.iter().enumerate() {
            let TokenKind::Identifier(name) = &token.kind else {
                continue;
            };
            if !Self::NAMES.contains(&name.to_lowercase().as_str()) {
                continue;
            }
            if index + 2 >= tokens.len() || tokens[index + 1].kind != TokenKind::LeftParen {
                continue;
            }

            let mut start = index + 2;
            let range = match &tokens[start].kind {
                TokenKind::Number(_) | TokenKind::String(_) => tokens[start].range.clone(),
                TokenKind::Identifier(word)
                    if word.eq_ignore_ascii_case("true") || word.eq_ignore_ascii_case("false") =>
                {
                    tokens[start].range.clone()
                }
                TokenKind::Minus => {
                    start += 1;
                    if start >= tokens.len() || !matches!(tokens[start].kind, TokenKind::Number(_))
                    {
                        return None;
                    }
                    tokens[start - 1].range.start..tokens[start].range.end
                }
                _ => return None,
            };

            let mut characters: Vec<char> = raw.chars().collect();
            characters.splice(range, literal.chars());
            return Some(characters.into_iter().collect());
        }
        None
    }

    /// Numeric convenience kept for slider drags (the Swift `Slider` enum).
    pub fn rewriting_value(raw: &str, new_value: &BigDecimal) -> Option<String> {
        Self::rewriting(raw, &new_value.to_string())
    }
}
